system/lib/wasmfs/file.h

```diff
@@ -261,6 +181,78 @@ class Symlink : public File {
   const std::string& getTarget() { return target; }
 };
 
+class File::Handle {
+protected:
+  // This mutex is needed when one needs to access access a previously locked
+  // file in the same thread. For example, rename will need to traverse
+  // 2 paths and access the same locked directory twice.
+  // TODO: During benchmarking, test recursive vs normal mutex performance.
+  std::unique_lock<std::recursive_mutex> lock;
+  std::shared_ptr<File> file;
+
+public:
+  Handle(std::shared_ptr<File> file) : file(file), lock(file->mutex) {}
+  Handle(std::shared_ptr<File> file, std::defer_lock_t)
+    : file(file), lock(file->mutex, std::defer_lock) {}
+  bool trylock() { return lock.try_lock(); }
+  size_t getSize() { return file->getSize(); }
+  mode_t& mode() { return file->mode; }
+  time_t& ctime() { return file->ctime; }
+  time_t& mtime() { return file->mtime; }
+  time_t& atime() { return file->atime; }
+
+  // Note: parent.lock() creates a new shared_ptr to the same Directory
+  // specified by the parent weak_ptr.
+  std::shared_ptr<File> getParent() { return file->parent.lock(); }
+  void setParent(std::shared_ptr<File> parent) { file->parent = parent; }
+
+  std::shared_ptr<File> unlocked() { return file; }
+};
+
+class DataFile::Handle : public File::Handle {
+  std::shared_ptr<DataFile> getFile() { return file->cast<DataFile>(); }
+
+public:
+  Handle(std::shared_ptr<File> dataFile) : File::Handle(dataFile) {}
+  Handle(Handle&&) = default;
+
+  __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) {
+    return getFile()->read(buf, len, offset);
+  }
+  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) {
+    return getFile()->write(buf, len, offset);
+  }
+
+  // TODO: Design a proper API for flushing files.
+  void flush() { getFile()->flush(); }
+
+  // This function loads preloaded files from JS Memory into this DataFile.
+  // TODO: Make this virtual so specific backends can specialize it for better
+  // performance.
+  void preloadFromJS(int index);
+};
+
+class Directory::Handle : public File::Handle {
+  std::shared_ptr<Directory> getDir() { return file->cast<Directory>(); }
+
+public:
+  Handle(std::shared_ptr<File> directory) : File::Handle(directory) {}
+  Handle(std::shared_ptr<File> directory, std::defer_lock_t)
+    : File::Handle(directory, std::defer_lock) {}
+
+  std::shared_ptr<File> getEntry(const std::string& name) {
+    return getDir()->getEntry(name);
+  }
+  std::shared_ptr<File> removeEntry(const std::string& name);
```

Comment:
the other `removeEntry` returns a boolean, while this one returns a shared File. I don't think I see the File used in any of the callers, so the non-matching API seems a little confusing and I'm not sure what the gain is from the divergence?

Reply:
Yeah, reasonable question. I'll change it to bool and we can change it back if we ever find we need to in the future.